pkg/controller/tas/topology_ungater.go

```diff
@@ -293,6 +301,67 @@ func podsToUngateInfo(
 }
 
 func assignGatedPodsToDomains(
+	log logr.Logger,
+	psa *kueue.PodSetAssignment,
+	pods []*corev1.Pod) []podWithDomain {
+	if rankToGatedPod, ok := readRanksIfAvailable(log, psa, pods); ok {
+		return assignGatedPodsToDomainsByRanks(psa, rankToGatedPod)
+	}
+	return assignGatedPodsToDomainsGreedy(log, psa, pods)
+}
+
+func assignGatedPodsToDomainsByRanks(
+	psa *kueue.PodSetAssignment,
+	rankToGatedPod map[int]*corev1.Pod) []podWithDomain {
+	toUngate := make([]podWithDomain, 0)
+	sortedDomains := sortDomains(psa)
+	totalCount := 0
+	for i := range sortedDomains {
+		totalCount += sortedDomains[i].count
+	}
+	rankToDomainID := make([]utiltas.TopologyDomainID, totalCount)
+	index := 0
+	for _, domain := range sortedDomains {
+		for s := range domain.count {
+			rankToDomainID[index+s] = domain.domainID
+		}
+		index += domain.count
+	}
+	for rank, pod := range rankToGatedPod {
+		toUngate = append(toUngate, podWithDomain{
+			pod:      pod,
+			domainID: rankToDomainID[rank],
+		})
+	}
+	return toUngate
+}
+
+func sortDomains(psa *kueue.PodSetAssignment) []domainWithCount {
+	type sortableDomainWithCount struct {
+		domainWithCount
+		sortName string
+	}
+	sortableDomains := make([]sortableDomainWithCount, len(psa.TopologyAssignment.Domains))
+	for i, domain := range psa.TopologyAssignment.Domains {
+		sortableDomains[i] = sortableDomainWithCount{
+			sortName: strings.Join(domain.Values, ","),
+			domainWithCount: domainWithCount{
+				domainID: utiltas.DomainID(domain.Values),
+				count:    int(domain.Count),
+			},
+		}
+	}
+	slices.SortFunc(sortableDomains, func(a, b sortableDomainWithCount) int {
+		return strings.Compare(a.sortName, b.sortName)
```

Comment:
How about joining `domain.Values` here to sort domains by it instead of introducing a new type

Reply:
We could techanically, but I think this would be worse performance, as then we need to recompute multiple times, when comparing say "a" against "b" then "c".

Reply:
Alternative is to simply make the TopologyDomainID type comparable.

Reply:
I have simplified it, PTAL

Reply:
Looks good, thank you